### ISSUE TITLE
cast, type/Type/Type[Any], and generics

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#2733]
**Describe the bug**
Not quite sure if this is a bug, but it's puzzling behavior. I am trying to store a type `Optional[x]` in a variable. In order to do this I am casting an upstream variable (`x`) to a type. I have tried casting to `type`, `Type` and `Type[Any]` and find the results baffling. I've annotated several lines with the output of LSP hover over the `cls` variable:

```python
from typing import cast, Type

def f(x: ):
    cls = cast(type, x)  # (variable) cls: type
    y = Optional[cls]  # (variable) cls: Unknown
    # errors:
    # - Illegal type annotation: variable not allowed unless it is a type alias
    # - Expected class type but received type

def g(x):
    cls = cast(Type, x)  # (variable) cls: Unknown
    y = Optional[cls]  # (variable) cls: Unknown
    # No errors!

def h(x):
    cls = cast(Type[Any], x)
    y = Optional[cls]
    # errors:
    # - Illegal type annotation: variable not allowed unless it is a type alias
    # - Expected class type but received type
```

Several questions:

1. Why does casting to bare `Type` behave differently from casting to `type` or `Type[Any]`? The [Python docs] say: "Type[Any] is equivalent to Type which in turn is equivalent to type, which is the root of Python’s metaclass hierarchy."
2. Why is variable `cls` `Unknown` in all cases inside the brackets? It's been cast on the preceding line.
3. Why do `f` and `h` give errors while `g` does not, despite the fact that `cls` is unknown for all of them?

**VS Code extension or command-line**
Pyright LSP 1.1.197

**Comment by erictraut** [comment]
The symbol `Optional` is a special form that is intended to be used within type annotations, but it appears that you're trying to use it as a runtime class. Is that correct? If so, why?

If your intent is to use `Optional` as a runtime class, you should use it without any subscript. If your intent is to use it as a type annotation, then you should use a subscript, but you shouldn't assign it to a variable.

The reason you're seeing an `Unknown` when you hover over `cls` in the expression `Optional[cls]` is because pyright is rejecting this use of `cls` because it doesn't follow the rules of the `Optional` special form, which is intended to be used only in type annotations, and `cls` is not a valid type argument.

I also don't understand why you are using a `cast` operator here. The expressions that you're casting have an `Any` type to begin with, so they are assignable to a variable of any type. Why not just declare the type of `cls` to be `type` if that is your intent? That is, use `cls: type = x`?
